lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,2561 @@
+  }
+
+  void replaceLeaf(VirtualQID old_qid, VirtualQID new_qid,
+                   DependencyNode *new_leaf) {
+    assert(new_leaf->isLeaf() && "Invalid leaf!");
+
+    if (leafs.count(old_qid) == 1) {
+      auto old_leaf = leafs[old_qid];
+
+      auto first_use = old_leaf->getSuccessorForQID(old_qid);
+      auto idx = first_use->getDependencyForQID(old_qid).value();
+
+      first_use->dependencies[idx] =
+          DependencyNode::DependencyEdge(new_leaf, 0);
+      old_leaf->successors.remove(first_use);
+      new_leaf->successors.insert(first_use);
+      if (old_leaf->isAlloc()) {
+        allocs.erase(allocs.find(old_qid));
+        auto alloc = static_cast<InitDependencyNode *>(old_leaf);
+        if (alloc->qubit)
+          qubits.erase(qubits.find(alloc->qubit.value()));
+      }
+    }
+
+    leafs[new_qid] = new_leaf;
+    if (new_leaf->isAlloc()) {
+      auto alloc = static_cast<InitDependencyNode *>(new_leaf);
+      allocs[new_qid] = alloc;
+      if (alloc->qubit)
+        qubits[alloc->qubit.value()] = alloc;
+    }
+  }
+
+  void replaceRoot(VirtualQID old_qid, VirtualQID new_qid,
+                   DependencyNode *new_root) {
+    assert(new_root->isRoot() && "Invalid root!");
+
+    if (qids.contains(old_qid)) {
+      auto old_root = getRootForQID(old_qid);
+
+      auto idx = old_root->getDependencyForQID(old_qid).value();
+
+      auto dep = old_root->dependencies[idx];
+      dep->successors.remove(old_root);
+      dep->successors.insert(new_root);
+
+      new_root->dependencies.push_back(dep);
+      old_root->dependencies.erase(old_root->dependencies.begin() + idx);
+
+      // If the terminator is somehow getting deleted, then the entire block
+      // must be empty, and then it will never be used
+      if (old_root->dependencies.size() == 0)
+        roots.remove(old_root);
+
+      old_root->qids.remove(old_qid);
+    }
+
+    new_root->qids.insert(new_qid);
+    // new_root->qids.insert(old_qid);
+    roots.insert(new_root);
+  }
+
+  /// Gathers all the nodes in the graph into seen, starting from next
+  void gatherNodes(SetVector<DependencyNode *> &seen, DependencyNode *next) {
+    if (seen.contains(next) || !next->isQuantumDependent())
+      return;
+
+    seen.insert(next);
+
+    for (auto successor : next->successors)
+      gatherNodes(seen, successor);
+    for (auto dependency : next->dependencies)
+      gatherNodes(seen, dependency.node);
+  }
+
+public:
+  DependencyGraph(DependencyNode *root) {
+    total_height = 0;
+    SetVector<DependencyNode *> seen;
+    qids = SetVector<VirtualQID>();
+    gatherRoots(seen, root);
+  }
+
+  ~DependencyGraph() {
+    SetVector<DependencyNode *> nodes;
+    for (auto root : roots)
+      gatherNodes(nodes, root);
+
+    for (auto node : nodes)
+      // ArgDependencyNodes are handled by the block and skipped here
+      if (!node->isLeaf() || !node->isQuantumOp() || node->isAlloc())
+        delete node;
```

Comment:
Where are the leaf nodes, quantum nodes and alloc nodes deleted?

Reply:
It's only quantum leaf nodes that are not allocs that are deleted here. As the comment says, `ArgDependencyNodes` are handled by the `DependencyBlock` destructor.